File: java-to-js/src/StreamCollectors.java
import java.util.List;

import static java.util.stream.Collectors.counting;
import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.mapping;
import static java.util.stream.Collectors.partitioningBy;
import static java.util.stream.Collectors.summarizingInt;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toSet;

record Student(String name, String grade, int score) {}

List<Student> students = List.of(
  new Student("Alice", "A", 90),
  new Student("Bob", "B", 85),
  new Student("Charlie", "C", 70),
  new Student("David", "A", 95),
  new Student("Eve", "B", 80)
);

void main() {
  var studentsByGrade = students.stream()
    .collect(groupingBy(Student::grade));

  var namesByGrade = students.stream()
    .collect(groupingBy(
      Student::grade,
      mapping(Student::name, toList())
    ));

  var countByGrade = students.stream()
    .collect(groupingBy(Student::grade, counting()));

  var allNames = students.stream()
    .map(Student::name)
    .collect(joining(", "));

  var uniqueGrades = students.stream()
    .map(Student::grade)
    .collect(toSet());

  var stats = students.stream()
    .collect(summarizingInt(Student::score));

  var awardStudents = students.stream()
    .collect(partitioningBy(student -> student.score() > 94));

  printResult("Students by grade", studentsByGrade);
  printResult("Names by grade", namesByGrade);
  printResult("Count by grade", countByGrade);
  printResult("All names", allNames);
  printResult("Unique grades", uniqueGrades);
  printResult("Stats", stats);
  printResult("Award Students", awardStudents);
}

void printResult(String title, Object args) {
  System.out.println(STR."--------------- \{title}---------------");
  System.out.println(args);
  System.out.println();
}

File: java-to-js/src/stream-collectors.ts
type Student = { name: string; grade: string; score: number };

const students: Student[] = [
  { name: "Alice", grade: "A", score: 90 },
  { name: "Bob", grade: "B", score: 85 },
  { name: "Charlie", grade: "C", score: 70 },
  { name: "David", grade: "A", score: 95 },
  { name: "Eve", grade: "B", score: 80 },
];

const studentsByGrade = students.reduce((acc, student) => {
  if (!acc.has(student.grade)) acc.set(student.grade, []);
  acc.get(student.grade)!.push(student);
  return acc;
}, new Map<string, Student[]>());
// Object.groupBy(students, (item) => item.grade);

const namesByGrade = students.reduce((acc, student) => {
  if (!acc.has(student.grade)) acc.set(student.grade, []);
  acc.get(student.grade)!.push(student.name);
  return acc;
}, new Map<string, string[]>());

const countByGrade = students.reduce((acc, student) => {
  const currentCount = acc.get(student.grade) || 0;
  acc.set(student.grade, currentCount + 1);
  return acc;
}, new Map<string, number>());

const allNames = students.map((student) => student.name).join(", ");

const uniqueGrades = students.reduce(
  (acc, student) => acc.add(student.grade),
  new Set<string>(),
);
// new Set(students.map((student) => student.grade));

const stats = students.reduce(
  (acc, student) => {
    acc.count++;
    acc.sum += student.score;
    acc.min = Math.min(acc.min, student.score);
    acc.average = acc.sum / acc.count;
    acc.max = Math.max(acc.max, student.score);
    return acc;
  },
  { count: 0, sum: 0, min: Infinity, average: 0, max: -Infinity },
);

const awardStudents = students.reduce((acc, student) => {
  const key = student.score > 94;
  if (!acc.has(key)) acc.set(key, []);
  acc.get(key)!.push(student);
  return acc;
}, new Map<boolean, Student[]>());

printResult("Students by grade", studentsByGrade);
printResult("Names by grade", namesByGrade);
printResult("Count by grade", countByGrade);
printResult("All names", allNames);
printResult("Unique grades", uniqueGrades);
printResult("Stats", stats);
printResult("Award Students", awardStudents);

function printResult(title: string, data: unknown) {
  console.log(`--------------- ${title} ---------------`);
  console.log(data);
  console.log();
}
